### PR TITLE
Add a codefix that suggests replacements for unknown identifiers

### PR DESCRIPTION
**Pull request by forki**
image

- [x] - Suggestions for FS0039
- [x] - Suggestions for other errors
- [x] - Put strings into resx

### CODE REVIEW

**Comment by forki** [comment]
@vasily-kirichenko could you please take a quick look if this goes into the right direction?

**Comment by forki** [comment]
this is ready for review

**Comment by vasily-kirichenko** [comment]
:)

image

**Comment by vasily-kirichenko** [comment]
LGTM

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            title)
+
+    override __.FixableDiagnosticIds = fixableDiagnosticIds.ToImmutableArray()
+
+    override __.RegisterCodeFixesAsync context : Task =
+       async { 
+            context.Diagnostics 
+            |> Seq.filter (fun x -> fixableDiagnosticIds |> List.contains x.Id)
+            |> Seq.iter (fun x ->
+                let splitted = x.GetMessage().Split([|"Maybe you want one of the following?"|], StringSplitOptions.None)
```
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: already fixed

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                if splitted.Length > 1 then
+                    let suggestions = 
+                        splitted.[1].Split([|' '; '\r'; '\n'|], StringSplitOptions.RemoveEmptyEntries) 
+                        |> Array.map (fun s -> s.Trim()) 
+                    let diagnostics = [| x |].ToImmutableArray()
+
+                    for suggestion in suggestions do
+                        let codefix = 
+                            createCodeFix(
+                                "Replace with " + suggestion,
```
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: already done

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                        splitted.[1].Split([|' '; '\r'; '\n'|], StringSplitOptions.RemoveEmptyEntries) 
+                        |> Array.map (fun s -> s.Trim()) 
+                    let diagnostics = [| x |].ToImmutableArray()
+
+                    for suggestion in suggestions do
+                        let codefix = 
+                            createCodeFix(
+                                "Replace with " + suggestion, 
+                                context, 
+                                TextChange(TextSpan(context.Span.Start, context.Span.End), suggestion))
```
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: already done

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+       async { 
+            context.Diagnostics 
+            |> Seq.filter (fun x -> fixableDiagnosticIds |> List.contains x.Id)
+            |> Seq.iter (fun x ->
+                let splitted = x.GetMessage().Split([|"Maybe you want one of the following?"|], StringSplitOptions.None)
+                if splitted.Length > 1 then
+                    let suggestions = 
+                        splitted.[1].Split([|' '; '\r'; '\n'|], StringSplitOptions.RemoveEmptyEntries) 
+                        |> Array.map (fun s -> s.Trim()) 
+                    let diagnostics = [| x |].ToImmutableArray()
```
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: > I think each code fix should contain all the diagnostic ids in its "fixing" list. Currently you generate a code fix for each id in context that this code fix supports.

what do I have to change?

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: I was wrong, forget it.

**Comment by vasily-kirichenko** [comment]
It still suggests unrelated stuff, like on the screenshot above. How on Earth `nativeptr` is close to `DateT`?

**Comment by forki** [comment]
It's close according to the given metric. There will always be false
positives...

Am 27.12.2016 15:29 schrieb "Vasily Kirichenko" <notifications@github.com>:

**Comment by isaacabraham** [comment]
Have we moved over to Jaro Winkler?

**Comment by forki** [comment]
@isaacabraham yes long time ago

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    override __.FixableDiagnosticIds = fixableDiagnosticIds.ToImmutableArray()
+
+    override __.RegisterCodeFixesAsync context : Task =
+        async { 
+            context.Diagnostics 
+            |> Seq.filter (fun x -> fixableDiagnosticIds |> Set.contains x.Id)
+            |> Seq.iter (fun diagnostic ->
+                let message = diagnostic.GetMessage()
+                let splitted = message.Split([|maybeString|], StringSplitOptions.None)
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: splitted is not really a word :-)

The past tense of split is .... split.

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                if splitted.Length > 1 then
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: This feels somewhat icky ... we should consider adding to the compiler service an improved method of transmitting error information to the IDE.

For example:

- I'm not at all sure how this string parsing we do in these fixers would work on a language like Arabic which works left to right.

Obviously that's not specific to this PR ... but we should consider an improvement.

Kevin

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs (hunk 5 above)
Comment (forki) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: https://en.m.wiktionary.org/wiki/splitted it says I'm nonstandard or archaic. Guess I can live with that. 

Anyways, will fix tomorrow.

**Pull request merged**